Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIPermissionsCocoa.mm

```diff
@@ -0,0 +1,124 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionAPIPermissions.h"
+
+#import "WebExtensionContextMessages.h"
+#import "WebProcess.h"
+#import <wtf/cocoa/VectorCocoa.h>
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+namespace WebKit {
+
+static NSString *permissionsKey = @"permissions";
+static NSString *originsKey = @"origins";
+
+void WebExtensionAPIPermissions::getAll(WebPage* page, Ref<WebExtensionCallbackHandler>&& callback)
+{
+    RELEASE_LOG(Extensions, "permissions.getAll()");
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::GetAll(page->webPageProxyIdentifier()), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](Vector<String> permissions, Vector<String> origins) {
+        callback->call(@{
+            permissionsKey: createNSArray(permissions).get(),
+            originsKey: createNSArray(origins).get()
+        });
+    }, extensionContext().identifier().toUInt64());
+}
+
+void WebExtensionAPIPermissions::contains(WebPage* page, NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **)
+{
+    RELEASE_LOG(Extensions, "permissions.contains()");
+
+    HashSet<String> permissions, origins;
+    parseDetailsDictionary(details, permissions, origins);
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::Contains(page->webPageProxyIdentifier(), permissions, origins), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](bool containsPermissions) {
+        callback->call(containsPermissions ? @YES : @NO);
+    }, extensionContext().identifier().toUInt64());
+}
+
+void WebExtensionAPIPermissions::request(WebPage* page, NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **)
+{
+    RELEASE_LOG(Extensions, "permissions.request()");
+
+    // FIXME: Check to see if the request is being made from a user gesture.
+
+    HashSet<String> permissions, origins;
+    parseDetailsDictionary(details, permissions, origins);
+
+    WebProcess::singleton().send(Messages::WebExtensionContext::Request(page->webPageProxyIdentifier(), permissions, origins), extensionContext().identifier());
+}
+
+void WebExtensionAPIPermissions::remove(WebPage* page, NSDictionary *details, Ref<WebExtensionCallbackHandler>&& callback, NSString **)
+{
+    RELEASE_LOG(Extensions, "permissions.remove()");
+
+    HashSet<String> permissions, origins;
+    parseDetailsDictionary(details, permissions, origins);
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::Remove(page->webPageProxyIdentifier(), permissions, origins), [ protectedThis = Ref { *this }, callback = WTFMove(callback) ](bool success, String errorMessage) {
```

Comment:
Drop the spaces inside the brackets: `[protectedThis = Ref { *this }, callback = WTFMove(callback)]`